File: training task/5_hybrid_inheritance.py
class Person:
    def __init__(self, name, age):
        self.name = name
        self.age = age

    def get_details(self):
        return f"Name: {self.name}, Age: {self.age}"

class Academic:
    def __init__(self, institution, major):
        self.institution = institution
        self.major = major

    def get_academic_details(self):
        return f"Institution: {self.institution}, Major: {self.major}"

class Student(Person, Academic):
    def __init__(self, name, age, institution, major, student_id):
        Person.__init__(self, name, age)
        Academic.__init__(self, institution, major)
        self.student_id = student_id

    def get_student_details(self):
        person_details = self.get_details()
        academic_details = self.get_academic_details()
        return f"{person_details}, {academic_details}, Student ID: {self.student_id}"

class GraduateStudent(Student):
    def __init__(self, name, age, institution, major, student_id, thesis_title):
        super().__init__(name, age, institution, major, student_id)
        self.thesis_title = thesis_title

    def get_graduate_student_details(self):
        student_details = self.get_student_details()
        return f"{student_details}, Thesis Title: {self.thesis_title}"

grad_student = GraduateStudent("Mou", 23, "NWU", "Computer Science", "G12345", "Machine Learning in Healthcare")
print(grad_student.get_graduate_student_details())

File: training task/5_hybrid_inheritance.cpp
#include <iostream>
#include <string>

using namespace std;
class Person {
private:
    string name;
    int age;

public:
    Person(string name, int age) {
        this->name = name;
        this->age = age;
    }
    string getDetails() const {
        return "Name: " +name + ", Age: " + to_string(age);
    }
};


class Academic {
protected:
    string institution;
    string major;

public:
    Academic(string institution, string major) {
        this->institution = institution;
        this->major = major;
    }

    
    string getAcademicDetails() const {
        return "Institution: " + institution + ", Major: " + major;
    }
};

class Student : public Person, public Academic {
private:
    string student_id;

public:
    Student(string name, int age, string institution, string major, string student_id)
        : Person(name, age), Academic(institution, major) {
        this->student_id = student_id;
    }

    // Method to get student details
    string getStudentDetails() const {
        string person_details = this->getDetails();
        string academic_details = this->getAcademicDetails();
        return person_details + ", " + academic_details + ", Student ID: " + student_id;
    }
};

class GraduateStudent : public Student {
private:
    string thesis_title;

public:
    GraduateStudent(string name, int age, string institution, string major, string student_id, string thesis_title)
        : Student(name, age, institution, major, student_id) {
        this->thesis_title = thesis_title;
    }

    string getGraduateStudentDetails() const {
        string student_details = getStudentDetails();
        return student_details + ", Thesis Title: " + thesis_title;
    }
};

int main() {
    GraduateStudent grad_student("Mou", 23, "NWU", "Computer Science", "G12345", "Machine Learning in Healthcare");

    cout << grad_student.getGraduateStudentDetails() << endl;

    return 0;
}
